File: modules/go/nyx/test/integration/git/util/git_util.go
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package util

import (
	"io"      // https://pkg.go.dev/io
	"strings" // https://pkg.go.dev/strings
)

/*
Prints a formatted separator with the given marker. This separator can be used to split various sections
in the output.

Arguments are as follows:

  - out the stream to print to
  - marker an optional string that can be used as a marker in the separator, to make
    it easier to jump to one dumo or the other when the output contains a lot of dumps.
    If empty a default marker is printed.
  - start prints different characters if this flag is true or false improving readability.
*/
func PrintSeparator(out io.Writer, marker string, start bool) {
	ch := ""
	if start {
		ch = ">"
	} else {
		ch = "<"
	}

	out.Write([]byte(strings.Repeat(ch, 10) + " "))
	if marker == "" {
		out.Write([]byte(strings.Repeat(ch, 20)))
	} else {
		out.Write([]byte(marker))
	}
	out.Write([]byte(" " + strings.Repeat(ch, 10)))
	out.Write([]byte("\n"))
}

/*
Dumps a whole lot of informations about a Git repository in the given directory to the given output stream.

Arguments are as follows:

  - directory the repository to dump the informations about
  - out the stream to write to
  - marker an optional string that can be used as a marker at the beginning and end of the dump, to make
    it easier to jump to one dumo or the other when the output contains a lot of dumps.
*/
func PrintRepositoryInfo(directory string, out io.Writer, marker string) {
	PrintSeparator(out, marker, true)

	PrintSeparator(out, "REPOSITORY FOLDER", true)
	RenderDirectoryTreeTo(directory, out)
	PrintSeparator(out, "REPOSITORY FOLDER", false)

	PrintSeparator(out, "REPOSITORY STATUS", true)
	PrintRepositoryStatus(directory, out)
	PrintSeparator(out, "REPOSITORY STATUS", false)

	PrintSeparator(out, "REPOSITORY OBJECTS", true)
	PrintRepositoryObjects(directory, out)
	PrintSeparator(out, "REPOSITORY OBJECTS", false)

	PrintSeparator(out, "REPOSITORY TAGS", true)
	PrintRepositoryTags(directory, out)
	PrintSeparator(out, "REPOSITORY TAGS", false)

	PrintSeparator(out, "REPOSITORY COMMIT HISTORY", true)
	PrintRepositoryCommitHistory(directory, out)
	PrintSeparator(out, "REPOSITORY COMMIT HISTORY", false)

	PrintSeparator(out, "REPOSITORY COMMIT GRAPH BY DATE", true)
	PrintRepositoryCommitGraphByDate(directory, out)
	PrintSeparator(out, "REPOSITORY COMMIT GRAPH BY DATE", false)

	PrintSeparator(out, "REPOSITORY COMMIT GRAPH BY TOPOLOGY", true)
	PrintRepositoryCommitGraphByTopology(directory, out)
	PrintSeparator(out, "REPOSITORY COMMIT GRAPH BY TOPOLOGY", false)

	PrintSeparator(out, marker, false)
}

/*
Dumps the repository status.

Arguments are as follows:

- directory the repository to dump the informations about
- out the stream to write to
*/
func PrintRepositoryStatus(directory string, out io.Writer) {
	RunCommand([]string{"git", "status"}, nil, &directory, out)
}

/*
Dumps the list of repository objects.

Arguments are as follows:

- directory the repository to dump the informations about
- out the stream to write to
*/
func PrintRepositoryObjects(directory string, out io.Writer) {
	RunCommand([]string{"git", "rev-list", "--objects", "--all"}, nil, &directory, out)
}

/*
Dumps the list of repository tags.

Arguments are as follows:

- directory the repository to dump the informations about
- out the stream to write to
*/
func PrintRepositoryTags(directory string, out io.Writer) {
	RunCommand([]string{"git", "tag", "--list", "-n"}, nil, &directory, out)
}

/*
Dumps the list of repository commits.

Arguments are as follows:

- directory the repository to dump the informations about
- out the stream to write to
*/
func PrintRepositoryCommitHistory(directory string, out io.Writer) {
	RunCommand([]string{"git", "log", "--decorate=full", "--source"}, nil, &directory, out)
}

/*
Dumps the list of repository commits with a graph sorted by date.

Arguments are as follows:

- directory the repository to dump the informations about
- out the stream to write to
*/
func PrintRepositoryCommitGraphByDate(directory string, out io.Writer) {
	RunCommand([]string{"git", "log", "--all", "--decorate", "--oneline", "--graph", "--date-order"}, nil, &directory, out)
}

/*
Dumps the list of repository commits with a graph sorted by topology.

Arguments are as follows:

- directory the repository to dump the informations about
- out the stream to write to
*/
func PrintRepositoryCommitGraphByTopology(directory string, out io.Writer) {
	RunCommand([]string{"git", "log", "--all", "--decorate", "--oneline", "--graph", "--topo-order"}, nil, &directory, out)
}

File: src/java/main/src/integrationTest/java/com/mooltiverse/oss/nyx/git/util/GitUtil.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.git.util;

import java.io.File;
import java.io.OutputStream;
import java.io.PrintStream;

import java.util.Objects;

/**
 * Git utilities used in tests
 */
public class GitUtil {
    /**
     * Default constructor is private on purpose.
     */
    private GitUtil() {
        super();
    }

    /**
     * Prints a formatted separator with the given marker. This separator can be used to split various sections
     * in the output.
     * 
     * @param out the stream to print to
     * @param marker an optional string that can be used as a marker in the separator, to make
     * it easier to jump to one dumo or the other when the output contains a lot of dumps.
     * If {@code null} a default marker is printed.
     * @param start prints different characters if this flag is {@code true} or {@code false}
     * improving readability.
     */
    private static void printSeparator(PrintStream out, String marker, boolean start) {
        String ch = start ? ">" : "<";
        out.print(ch.repeat(10)+" ");
        out.print(Objects.isNull(marker) ? ch.repeat(20) : marker);
        out.print(" "+ch.repeat(10));
        out.println();
    }

    /**
     * Dumps a whole lot of informations about a Git repository in the given directory to the given output stream.
     * 
     * @param directory the repository to dump the informations about
     * @param out the stream to write to
     * @param marker an optional string that can be used as a marker at the beginning and end of the dump, to make
     * it easier to jump to one dumo or the other when the output contains a lot of dumps.
     */
    public static void printRepositoryInfo(File directory, OutputStream out, String marker)
        throws Exception {
        PrintStream ps = new PrintStream(out);
        printSeparator(ps, marker, true);

        printSeparator(ps, "REPOSITORY FOLDER", true);
        FileSystemUtil.renderDirectoryTree(directory, out);
        printSeparator(ps, "REPOSITORY FOLDER", false);

        printSeparator(ps, "REPOSITORY STATUS", true);
        printRepositoryStatus(directory, out);
        printSeparator(ps, "REPOSITORY STATUS", false);

        printSeparator(ps, "REPOSITORY OBJECTS", true);
        printRepositoryObjects(directory, out);
        printSeparator(ps, "REPOSITORY OBJECTS", false);

        printSeparator(ps, "REPOSITORY TAGS", true);
        printRepositoryTags(directory, out);
        printSeparator(ps, "REPOSITORY TAGS", false);

        printSeparator(ps, "REPOSITORY COMMIT HISTORY", true);
        printRepositoryCommitHistory(directory, out);
        printSeparator(ps, "REPOSITORY COMMIT HISTORY", false);

        printSeparator(ps, "REPOSITORY COMMIT GRAPH BY DATE", true);
        printRepositoryCommitGraphByDate(directory, out);
        printSeparator(ps, "REPOSITORY COMMIT GRAPH BY DATE", false);

        printSeparator(ps, "REPOSITORY COMMIT GRAPH BY TOPOLOGY", true);
        printRepositoryCommitGraphByTopology(directory, out);
        printSeparator(ps, "REPOSITORY COMMIT GRAPH BY TOPOLOGY", false);

        printSeparator(ps, marker, false);
        ps.flush();
    }

    /**
     * Dumps the repository status.
     * 
     * @param directory the repository to dump the informations about
     * @param out the stream to write to
     */
    public static void printRepositoryStatus(File directory, OutputStream out)
        throws Exception {
        RuntimeUtil.runCommand(new String[]{"git", "status"}, null, directory, out);
    }

    /**
     * Dumps the list of repository objects.
     * 
     * @param directory the repository to dump the informations about
     * @param out the stream to write to
     */
    public static void printRepositoryObjects(File directory, OutputStream out)
        throws Exception {
        RuntimeUtil.runCommand(new String[]{"git", "rev-list", "--objects", "--all"}, null, directory, out);
    }

    /**
     * Dumps the list of repository tags.
     * 
     * @param directory the repository to dump the informations about
     * @param out the stream to write to
     */
    public static void printRepositoryTags(File directory, OutputStream out)
        throws Exception {
        RuntimeUtil.runCommand(new String[]{"git", "tag", "--list", "-n"}, null, directory, out);
    }

    /**
     * Dumps the list of repository commits.
     * 
     * @param directory the repository to dump the informations about
     * @param out the stream to write to
     */
    public static void printRepositoryCommitHistory(File directory, OutputStream out)
        throws Exception {
        RuntimeUtil.runCommand(new String[]{"git", "log", "--decorate=full", "--source"}, null, directory, out);
    }

    /**
     * Dumps the list of repository commits with a graph sorted by date.
     * 
     * @param directory the repository to dump the informations about
     * @param out the stream to write to
     */
    public static void printRepositoryCommitGraphByDate(File directory, OutputStream out)
        throws Exception {
        RuntimeUtil.runCommand(new String[]{"git", "log", "--all", "--decorate", "--oneline", "--graph", "--date-order"}, null, directory, out);
    }

    /**
     * Dumps the list of repository commits with a graph sorted by topology.
     * 
     * @param directory the repository to dump the informations about
     * @param out the stream to write to
     */
    public static void printRepositoryCommitGraphByTopology(File directory, OutputStream out)
        throws Exception {
        RuntimeUtil.runCommand(new String[]{"git", "log", "--all", "--decorate", "--oneline", "--graph", "--topo-order"}, null, directory, out);
    }
}